excel/src/main/java/org/apache/metamodel/excel/DefaultSpreadsheetReaderDelegate.java

```diff
@@ -146,34 +157,196 @@ private MutableTable createTable(final Workbook wb, final Sheet sheet) {
+        int numberOfLinesToScan = _configuration.getNumberOfLinesToScan() - 1;
+
+        while (dataRowIterator.hasNext() && numberOfLinesToScan-- > 0) {
+            final Row currentRow = dataRowIterator.next();
+            if (currentRow != null && currentRow.getLastCellNum() > 0) {
+                for (int i = getColumnOffset(currentRow); i < columnTypes.length; i++) {
+                    final ColumnType detectNewColumnType = detectNewColumnTypeCell(columnTypes[i], currentRow
+                            .getCell(i));
+                    if (detectNewColumnType != null) {
+                        columnTypes[i] = detectNewColumnType;
+                    }
+                }
             }
         }
+    }
 
-        return table;
+    /**
+     * Tries to detect a new {@link ColumnType} for a cell.
+     * @param currentColumnType
+     * @param cell
+     * @return Returns a new {@link ColumnType} when detected. Otherwise null is returned.
+     */
+    private static ColumnType detectNewColumnTypeCell(final ColumnType currentColumnType, final Cell cell) {
+        // Can't detect something new if it's already on the default.
+        if (currentColumnType != null && currentColumnType.equals(DEFAULT_COLUMN_TYPE)) {
+            return null;
+        }
+        // Skip if the cell is null. This way 1 missing cell can't influence the column type of all other cells.
+        if (cell == null) {
+            return null;
+        }
+
+        final ColumnType detectedColumnType = determineColumnTypeFromCell(cell);
+        if (currentColumnType == null) {
+            return detectedColumnType;
+        } else if (!currentColumnType.equals(detectedColumnType)) {
+            // If the column type is Double and a Integer is detected, then don't set it to Integer
+            if (currentColumnType.equals(ColumnType.INTEGER) && detectedColumnType.equals(ColumnType.DOUBLE)) {
+                // If the column type is Integer and a Double is detected, then set it to Double
+                return detectedColumnType;
+            } else if (currentColumnType.equals(ColumnType.DOUBLE) && detectedColumnType.equals(ColumnType.INTEGER)) {
+                return null;
+            } else {
+                return DEFAULT_COLUMN_TYPE;
+            }
+        }
+        return null;
+    }
+
+    private static ColumnType determineColumnTypeFromCell(final Cell cell) {
+        switch (cell.getCellType()) {
+        case NUMERIC:
+            if (DateUtil.isCellDateFormatted(cell)) {
+                return ColumnType.DATE;
+            } else {
+                return cell.getNumericCellValue() % 1 == 0 ? ColumnType.INTEGER : ColumnType.DOUBLE;
+            }
+        case BOOLEAN:
+            return ColumnType.BOOLEAN;
+        case FORMULA:
+            final FormulaEvaluator evaluator = cell
+            .getSheet()
+            .getWorkbook()
+            .getCreationHelper()
+            .createFormulaEvaluator();
```

Comment:
These lines are formatted in a really weird manner. Next to that, I think that for performance reasons, it's better to create the `FormulaEvaluator` only once and then reuse it, since it's tied to the `Workbook` and not `Cell` specific.

Reply:
done, turned it into a class member, not sure if that goes to far?

Reply:
I think that goes to far, because the FormulaEvaluator is tied to the Workbook, which itself is not stored as a member class.

Reply:
ok, I get the workbook from the cell now

Reply:
Now we're back to the situation where I first commented on. I think there are three options to solve this:

1. Introduce a `LazyRef<FormulaEvaluator>` variable which you declare, for example, in `detectColumnTypes(Row, Iterator<Row>, ColumnType[] columnTypes`, and then pass on to the methods invoked by that method, which gets the FormulaEvaluator only the first time it's needed and reuses that for the rest of the process.
2. Introduce an inner class called `ColumnTypeScanner` (or something similar), which you instantiate on line 137 (inside the `createTable(Workbook, Sheet)` method and where you pass the Sheet into the constructor. This `ColumnTypeScanner` then contains all column type scanner related methods and also has a `FormulaEvaluator` field, which is set once the `FormulaEvaluator` is used for the first time, and after that it can be reused.
3. Same as option 2, but instead of using an inner class, create a normal class. This makes reusing some of the `DefaultSpreadsheetReaderDelegate` class's methods a bit harder.

Reply:
I like the ColumnTypeScanner as Inner Class. I've positioned the instantion in a different place, but I think you'll probably like this better too :).